File: java_solutions_1/433.minimum-genetic-mutation.java
/*
 * @lc app=leetcode id=433 lang=java
 *
 * [433] Minimum Genetic Mutation
 */

class Solution {
    public int minMutation(String startGene, String endGene, String[] bank) {
        return java.util.stream.Stream.of(0).map(_v -> new Object[]{new java.util.HashSet<>(java.util.Arrays.asList(bank)), new java.util.HashSet<String>(), new java.util.LinkedList<String>()}).peek(a -> ((java.util.Queue<String>) a[2]).offer(startGene)).peek(a -> ((java.util.Set<String>) a[1]).add(startGene)).map(a -> java.util.stream.Stream.iterate(0, steps -> !((java.util.Queue<String>) a[2]).isEmpty(), steps -> steps + 1).filter(steps -> java.util.stream.IntStream.range(0, ((java.util.Queue<String>) a[2]).size()).mapToObj(_i -> ((java.util.Queue<String>) a[2]).poll()).anyMatch(curr -> curr.equals(endGene) || java.util.stream.Stream.of(0).peek(_v -> java.util.stream.IntStream.range(0, 8).forEach(j -> java.util.Arrays.asList('A', 'C', 'G', 'T').forEach(cChar -> java.util.stream.Stream.of(curr.substring(0, j) + cChar + curr.substring(j + 1)).filter(mut -> ((java.util.Set<String>) a[0]).contains(mut) && ((java.util.Set<String>) a[1]).add(mut)).forEach(((java.util.Queue<String>) a[2])::offer)))).anyMatch(_v -> false))).findFirst().orElse(-1)).findFirst().get();
    }
}

File: kotlin_solutions/433.minimum-genetic-mutation.kt
/*
 * @lc app=leetcode id=433 lang=kotlin
 *
 * [433] Minimum Genetic Mutation
 */

class Solution { fun minMutation(startGene: String, endGene: String, bank: Array<String?>): Int = intArrayOf(-1).let { found -> generateSequence(LinkedList<Pair<String, Int>>().apply { add(startGene to 0) } to bank) { (q, b) -> q.poll()?.let { (curr, steps) -> b.indices.fold(q to b) { (queue, bankArray), i -> bankArray[i]?.let { next -> if (curr.indices.count { curr[it] != next[it] } == 1) { if (next == endGene) (steps + 1).also { found[0] = it }.let { queue to bankArray } else (queue.apply { add(next to (steps + 1)) }) to bankArray.apply { this[i] = null } } else queue to bankArray } ?: (queue to bankArray) }.takeIf { found[0] == -1 } } }.takeWhile { found[0] == -1 }.count().let { found[0] } } }

